flake.nix

```
@@ -0,0 +1,83 @@
+{
+  description = "Cyberus Guest Tests";
+
+  inputs = {
+    flake-parts.url = "github:hercules-ci/flake-parts";
+    flake-parts.inputs.nixpkgs-lib.follows = "nixpkgs";
+    nixpkgs.url = "github:nixos/nixpkgs/nixos-24.11";
+    nixpkgs-23-11.url = "github:nixos/nixpkgs/nixos-23.11";
+    pre-commit-hooks.url = "github:cachix/git-hooks.nix";
+    pre-commit-hooks.inputs.nixpkgs.follows = "nixpkgs";
+  };
+
+  outputs =
+    { self, flake-parts, ... }@inputs:
+    flake-parts.lib.mkFlake { inherit inputs; } {
+      flake = {
+        # non per-system flake attributes
+      };
+      # No artificially limit here. If it builds, it builds.
+      systems = inputs.nixpkgs.lib.systems.flakeExposed;
+      perSystem =
+        {
+          pkgs,
+          lib,
+          system,
+          ...
+        }:
+        let
+          project = import ./nix/release.nix {
+            inherit pkgs;
+            pkgs-23-11 = import inputs.nixpkgs-23-11 { inherit system; };
+          };
+        in
+        {
+          checks = {
+            pre-commit = import ./nix/pre-commit-check.nix {
+              pre-commit-hooks = inputs.pre-commit-hooks.lib.${system};
+              inherit pkgs;
+            };
+          };
+          devShells = {
+            # Shell for this repository.
+            default = pkgs.mkShell {
+              inputsFrom = builtins.attrValues self.packages.${system};
+              packages =
+                with pkgs;
+                [
+                  ninja
+                ]
+                ++ self.checks.${system}.pre-commit.enabledPackages;
+              shellHook = ''
+                ${self.checks.${system}.pre-commit.shellHook}
+              '';
+            };
+          };
+          formatter = pkgs.nixfmt-rfc-style;
+          packages =
+            let
+              # Unflattens a nested attribute set recursively, by constructing a
+              # new flat attribute key from the whole path.
+              flattenAttrsByName =
+                attrSet: prefix: sep:
+                builtins.foldl' (
+                  acc: key:
+                  let
+                    fullKey = if prefix == "" then key else "${prefix}${sep}${key}";
+                    value = attrSet.${key};
+                  in
+                  if !lib.isDerivation value then
+                    acc // (flattenAttrsByName value fullKey sep)
+                  else
+                    acc // { ${fullKey} = value; }
+                ) { } (builtins.attrNames attrSet);
```

I think we can improve this and remove manual handling of accumulators necessary for `foldl'` to work:

```suggestion
              # Flattens a nested attribute set recursively, by constructing a
              # new flat attribute key from the whole path.
              flattenAttrsByName =
                attrSet: prefix: sep:
                lib.attrsets.concatMapAttrs
                  (key: value:
                  let
                    fullKey = "${prefix}${sep}${key}";
                  in
                  if !lib.isDerivation value then
                    flattenAttrsByName value fullKey sep
                  else
                    { ${fullKey} = value; }
                  ) attrSet;
```

Note the fixed comment in line 59.

Also we might want to check for `lib.isAttrs value` in addition to `!lib.isDerivation`, so we don't trip over values of other types in the nested attrsets.

---

Thanks! That's helpful!